File: plugins/agentic-workflows/src/agentic_workflows/progress.py
"""Workflow progress tracking."""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Any

from filelock import FileLock


class WorkflowStatus(Enum):
    """Workflow execution status."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    PAUSED = "paused"
    CANCELLED = "cancelled"


class StepStatus(Enum):
    """Step execution status."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    SKIPPED = "skipped"


@dataclass
class StepProgress:
    """Progress information for a single step."""

    name: str
    status: str
    started_at: str | None = None
    completed_at: str | None = None
    retry_count: int = 0
    output_summary: str = ""
    error: str | None = None
    human_input: str | None = None


@dataclass
class ParallelBranch:
    """Information about a parallel branch execution."""

    branch_id: str
    status: str
    worktree_path: str
    progress_file: str


@dataclass
class WorkflowProgress:
    """Complete workflow progress state."""

    schema_version: str = "1.0"
    workflow_id: str = ""
    workflow_name: str = ""
    status: str = "pending"
    started_at: str | None = None
    completed_at: str | None = None
    current_step: dict[str, Any] | None = None
    completed_steps: list[StepProgress] = field(default_factory=list)
    pending_steps: list[str] = field(default_factory=list)
    running_steps: list[str] = field(default_factory=list)
    parallel_branches: list[ParallelBranch] = field(default_factory=list)
    errors: list[dict[str, Any]] = field(default_factory=list)
    variables: dict[str, Any] = field(default_factory=dict)
    step_outputs: dict[str, Any] = field(default_factory=dict)


def get_progress_path(workflow_id: str, repo_root: Path | None = None) -> Path:
    """Get path to progress file for a workflow."""
    if repo_root is None:
        repo_root = Path.cwd()
    return repo_root / "agentic" / "workflows" / workflow_id / "progress.json"


def load_progress(workflow_id: str, repo_root: Path | None = None) -> WorkflowProgress | None:
    """Load workflow progress from file."""
    progress_path = get_progress_path(workflow_id, repo_root)
    if not progress_path.exists():
        return None

    with open(progress_path, encoding="utf-8") as f:
        data = json.load(f)

    return _dict_to_progress(data)


def save_progress(progress: WorkflowProgress, repo_root: Path | None = None) -> None:
    """Save workflow progress to file with file locking."""
    progress_path = get_progress_path(progress.workflow_id, repo_root)
    progress_path.parent.mkdir(parents=True, exist_ok=True)

    lock_path = progress_path.with_suffix(".lock")
    lock = FileLock(str(lock_path))

    with lock, open(progress_path, "w", encoding="utf-8") as f:
        json.dump(_progress_to_dict(progress), f, indent=2)


def create_progress(
    workflow_id: str,
    workflow_name: str,
    step_names: list[str],
    variables: dict[str, Any],
) -> WorkflowProgress:
    """Create initial progress document."""
    return WorkflowProgress(
        workflow_id=workflow_id,
        workflow_name=workflow_name,
        status=WorkflowStatus.RUNNING.value,
        started_at=datetime.now(timezone.utc).isoformat(),
        pending_steps=step_names.copy(),
        variables=variables.copy(),
    )


def update_step_started(progress: WorkflowProgress, step_name: str) -> None:
    """Mark a step as started."""
    if step_name in progress.pending_steps:
        progress.pending_steps.remove(step_name)
    if step_name not in progress.running_steps:
        progress.running_steps.append(step_name)
    progress.current_step = {
        "name": step_name,
        "retry_count": 0,
        "started_at": datetime.now(timezone.utc).isoformat(),
    }


def update_step_completed(
    progress: WorkflowProgress,
    step_name: str,
    output_summary: str = "",
    output: Any = None,
) -> None:
    """Mark a step as completed."""
    if step_name in progress.running_steps:
        progress.running_steps.remove(step_name)

    started_at = None
    retry_count = 0
    if progress.current_step:
        started_at = progress.current_step.get("started_at")
        retry_count = progress.current_step.get("retry_count", 0)

    step = StepProgress(
        name=step_name,
        status=StepStatus.COMPLETED.value,
        started_at=started_at,
        completed_at=datetime.now(timezone.utc).isoformat(),
        retry_count=retry_count,
        output_summary=output_summary,
    )
    progress.completed_steps.append(step)

    if output is not None:
        progress.step_outputs[step_name] = output

    progress.current_step = None


def update_step_failed(
    progress: WorkflowProgress,
    step_name: str,
    error: str,
) -> None:
    """Mark a step as failed."""
    if step_name in progress.running_steps:
        progress.running_steps.remove(step_name)

    started_at = None
    retry_count = 0
    if progress.current_step:
        started_at = progress.current_step.get("started_at")
        retry_count = progress.current_step.get("retry_count", 0)

    step = StepProgress(
        name=step_name,
        status=StepStatus.FAILED.value,
        started_at=started_at,
        completed_at=datetime.now(timezone.utc).isoformat(),
        retry_count=retry_count,
        error=error,
    )
    progress.completed_steps.append(step)
    progress.errors.append({"step": step_name, "error": error})
    progress.current_step = None


def update_step_skipped(progress: WorkflowProgress, step_name: str) -> None:
    """Mark a step as skipped."""
    if step_name in progress.pending_steps:
        progress.pending_steps.remove(step_name)
    if step_name in progress.running_steps:
        progress.running_steps.remove(step_name)

    step = StepProgress(
        name=step_name,
        status=StepStatus.SKIPPED.value,
        completed_at=datetime.now(timezone.utc).isoformat(),
    )
    progress.completed_steps.append(step)
    progress.current_step = None


def _progress_to_dict(progress: WorkflowProgress) -> dict[str, Any]:
    """Convert progress to dictionary for JSON serialization."""
    data = asdict(progress)
    data["completed_steps"] = [asdict(s) if isinstance(s, StepProgress) else s for s in progress.completed_steps]
    data["parallel_branches"] = [asdict(b) if isinstance(b, ParallelBranch) else b for b in progress.parallel_branches]
    return data


def _dict_to_progress(data: dict[str, Any]) -> WorkflowProgress:
    """Convert dictionary to WorkflowProgress."""
    completed = [StepProgress(**s) if isinstance(s, dict) else s for s in data.get("completed_steps", [])]
    branches = [ParallelBranch(**b) if isinstance(b, dict) else b for b in data.get("parallel_branches", [])]

    return WorkflowProgress(
        schema_version=data.get("schema_version", "1.0"),
        workflow_id=data.get("workflow_id", ""),
        workflow_name=data.get("workflow_name", ""),
        status=data.get("status", "pending"),
        started_at=data.get("started_at"),
        completed_at=data.get("completed_at"),
        current_step=data.get("current_step"),
        completed_steps=completed,
        pending_steps=data.get("pending_steps", []),
        running_steps=data.get("running_steps", []),
        parallel_branches=branches,
        errors=data.get("errors", []),
        variables=data.get("variables", {}),
        step_outputs=data.get("step_outputs", {}),
    )

File: src/progress.ts
/** Workflow progress tracking. */

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { lockSync } from "proper-lockfile";
import type { CurrentStepInfo, ParallelBranch, StepProgress, WorkflowProgress } from "./types.js";

// --- Constants ---

export const WORKFLOW_STATUS = {
	PENDING: "pending",
	RUNNING: "running",
	COMPLETED: "completed",
	FAILED: "failed",
	PAUSED: "paused",
	CANCELED: "canceled",
} as const;

export const STEP_STATUS = {
	PENDING: "pending",
	RUNNING: "running",
	COMPLETED: "completed",
	FAILED: "failed",
	SKIPPED: "skipped",
} as const;

// --- ID and path helpers ---

export function generateWorkflowId(workflowName: string): string {
	const now = new Date();
	const pad = (n: number, len = 2) => String(n).padStart(len, "0");
	const timestamp = `${now.getUTCFullYear()}${pad(now.getUTCMonth() + 1)}${pad(now.getUTCDate())}-${pad(now.getUTCHours())}${pad(now.getUTCMinutes())}${pad(now.getUTCSeconds())}`;

	let safeName = workflowName.toLowerCase().replace(/ /g, "-").replace(/_/g, "-");
	safeName = safeName.replace(/[^a-z0-9-]/g, "");
	return `${timestamp}-${safeName}`;
}

export function getProgressPath(workflowId: string, outputDir: string): string {
	return path.join(outputDir, "progress.json");
}

// --- CRUD operations ---

export function loadProgress(workflowId: string, outputDir: string): WorkflowProgress | null {
	const progressPath = getProgressPath(workflowId, outputDir);
	if (!existsSync(progressPath)) {
		return null;
	}
	try {
		const data = JSON.parse(readFileSync(progressPath, "utf-8"));
		return dictToProgress(data);
	} catch {
		return null;
	}
}

export function saveProgress(progress: WorkflowProgress, outputDir: string): void {
	const progressPath = getProgressPath(progress.workflowId, outputDir);
	const dir = path.dirname(progressPath);
	mkdirSync(dir, { recursive: true });

	// Ensure the file exists so proper-lockfile can resolve its path
	if (!existsSync(progressPath)) {
		writeFileSync(progressPath, "{}", "utf-8");
	}

	let release: (() => void) | undefined;
	try {
		release = lockSync(progressPath);
		writeFileSync(progressPath, JSON.stringify(progressToDict(progress), null, 2), "utf-8");
	} finally {
		if (release) {
			release();
		}
	}
}

export function createProgress(
	workflowId: string,
	workflowName: string,
	stepNames: string[],
	variables: Record<string, unknown>,
	workflowFile = "",
): WorkflowProgress {
	return {
		schemaVersion: "1.0",
		workflowId,
		workflowName,
		status: WORKFLOW_STATUS.RUNNING,
		startedAt: new Date().toISOString(),
		completedAt: null,
		currentStep: null,
		completedSteps: [],
		pendingSteps: [...stepNames],
		runningSteps: [],
		parallelBranches: [],
		errors: [],
		variables: { ...variables },
		stepOutputs: {},
		workflowFile,
	};
}

// --- Step update functions ---

export function updateStepStarted(progress: WorkflowProgress, stepName: string): void {
	const pendingIdx = progress.pendingSteps.indexOf(stepName);
	if (pendingIdx !== -1) {
		progress.pendingSteps.splice(pendingIdx, 1);
	}
	if (!progress.runningSteps.includes(stepName)) {
		progress.runningSteps.push(stepName);
	}
	progress.currentStep = {
		name: stepName,
		retryCount: 0,
		startedAt: new Date().toISOString(),
	};
}

export function updateStepCompleted(
	progress: WorkflowProgress,
	stepName: string,
	outputSummary = "",
	output?: unknown,
): void {
	const runIdx = progress.runningSteps.indexOf(stepName);
	if (runIdx !== -1) {
		progress.runningSteps.splice(runIdx, 1);
	}

	const step: StepProgress = {
		name: stepName,
		status: STEP_STATUS.COMPLETED,
		startedAt: progress.currentStep?.startedAt ?? null,
		completedAt: new Date().toISOString(),
		retryCount: progress.currentStep?.retryCount ?? 0,
		outputSummary,
		error: null,
		humanInput: null,
	};
	progress.completedSteps.push(step);

	if (output !== undefined) {
		progress.stepOutputs[stepName] = output;
	}

	progress.currentStep = null;
}

export function updateStepFailed(
	progress: WorkflowProgress,
	stepName: string,
	error: string,
): void {
	const runIdx = progress.runningSteps.indexOf(stepName);
	if (runIdx !== -1) {
		progress.runningSteps.splice(runIdx, 1);
	}

	const step: StepProgress = {
		name: stepName,
		status: STEP_STATUS.FAILED,
		startedAt: progress.currentStep?.startedAt ?? null,
		completedAt: new Date().toISOString(),
		retryCount: progress.currentStep?.retryCount ?? 0,
		error,
		outputSummary: "",
		humanInput: null,
	};
	progress.completedSteps.push(step);
	progress.errors.push({ step: stepName, error });
	progress.currentStep = null;
}

export function updateStepSkipped(progress: WorkflowProgress, stepName: string): void {
	const pendingIdx = progress.pendingSteps.indexOf(stepName);
	if (pendingIdx !== -1) {
		progress.pendingSteps.splice(pendingIdx, 1);
	}
	const runIdx = progress.runningSteps.indexOf(stepName);
	if (runIdx !== -1) {
		progress.runningSteps.splice(runIdx, 1);
	}

	const step: StepProgress = {
		name: stepName,
		status: STEP_STATUS.SKIPPED,
		startedAt: null,
		completedAt: new Date().toISOString(),
		retryCount: 0,
		outputSummary: "",
		error: null,
		humanInput: null,
	};
	progress.completedSteps.push(step);
	progress.currentStep = null;
}

export function prepareForResume(progress: WorkflowProgress): void {
	if (progress.runningSteps.length > 0) {
		progress.pendingSteps = [...progress.runningSteps, ...progress.pendingSteps];
		progress.runningSteps = [];
	}

	if (progress.currentStep) {
		const stepName = progress.currentStep.name;
		if (stepName && !progress.pendingSteps.includes(stepName)) {
			progress.pendingSteps.unshift(stepName);
		}
		// Preserve humanInput so the resumed step can read the response
		if (!progress.currentStep.humanInput) {
			progress.currentStep = null;
		}
	}

	progress.status = WORKFLOW_STATUS.RUNNING;
	progress.completedAt = null;
}

// --- Serialization helpers ---

export function progressToDict(progress: WorkflowProgress): Record<string, unknown> {
	return {
		schema_version: progress.schemaVersion,
		workflow_id: progress.workflowId,
		workflow_name: progress.workflowName,
		status: progress.status,
		started_at: progress.startedAt,
		completed_at: progress.completedAt,
		current_step: progress.currentStep
			? {
					name: progress.currentStep.name,
					retry_count: progress.currentStep.retryCount,
					started_at: progress.currentStep.startedAt,
					type: progress.currentStep.type,
					message: progress.currentStep.message,
					timeout_minutes: progress.currentStep.timeoutMinutes,
					on_timeout: progress.currentStep.onTimeout,
					human_input: progress.currentStep.humanInput,
				}
			: null,
		completed_steps: progress.completedSteps.map((s) => ({
			name: s.name,
			status: s.status,
			started_at: s.startedAt,
			completed_at: s.completedAt,
			retry_count: s.retryCount,
			output_summary: s.outputSummary,
			error: s.error,
			human_input: s.humanInput,
		})),
		pending_steps: progress.pendingSteps,
		running_steps: progress.runningSteps,
		parallel_branches: progress.parallelBranches.map((b) => ({
			branch_id: b.branchId,
			status: b.status,
			worktree_path: b.worktreePath,
			progress_file: b.progressFile,
		})),
		errors: progress.errors,
		variables: progress.variables,
		step_outputs: progress.stepOutputs,
		workflow_file: progress.workflowFile,
	};
}

export function dictToProgress(data: Record<string, unknown>): WorkflowProgress {
	const completedSteps = ((data.completed_steps as Record<string, unknown>[]) ?? []).map(
		(s): StepProgress => ({
			name: (s.name as string) ?? "",
			status: (s.status as string) ?? "pending",
			startedAt: (s.started_at as string) ?? null,
			completedAt: (s.completed_at as string) ?? null,
			retryCount: (s.retry_count as number) ?? 0,
			outputSummary: (s.output_summary as string) ?? "",
			error: (s.error as string) ?? null,
			humanInput: (s.human_input as string) ?? null,
		}),
	);

	const parallelBranches = ((data.parallel_branches as Record<string, unknown>[]) ?? []).map(
		(b): ParallelBranch => ({
			branchId: (b.branch_id as string) ?? "",
			status: (b.status as string) ?? "",
			worktreePath: (b.worktree_path as string) ?? "",
			progressFile: (b.progress_file as string) ?? "",
		}),
	);

	return {
		schemaVersion: (data.schema_version as string) ?? "1.0",
		workflowId: (data.workflow_id as string) ?? "",
		workflowName: (data.workflow_name as string) ?? "",
		status: (data.status as string) ?? "pending",
		startedAt: (data.started_at as string) ?? null,
		completedAt: (data.completed_at as string) ?? null,
		currentStep: data.current_step
			? (() => {
					const cs = data.current_step as Record<string, unknown>;
					return {
						name: (cs.name as string) ?? "",
						retryCount: (cs.retry_count as number) ?? 0,
						startedAt: (cs.started_at as string) ?? "",
						type: cs.type as string | undefined,
						message: (cs.message as string | null) ?? null,
						timeoutMinutes: cs.timeout_minutes as number | undefined,
						onTimeout: cs.on_timeout as string | undefined,
						humanInput: (cs.human_input as string | null) ?? null,
					} satisfies CurrentStepInfo;
				})()
			: null,
		completedSteps,
		pendingSteps: (data.pending_steps as string[]) ?? [],
		runningSteps: (data.running_steps as string[]) ?? [],
		parallelBranches,
		errors: (data.errors as Record<string, unknown>[]) ?? [],
		variables: (data.variables as Record<string, unknown>) ?? {},
		stepOutputs: (data.step_outputs as Record<string, unknown>) ?? {},
		workflowFile: (data.workflow_file as string) ?? "",
	};
}
